go/service/kvstore.go

```diff
@@ -237,3 +242,182 @@ func (h *KVStoreHandler) PutKVEntry(ctx context.Context, arg keybase1.PutKVEntry
+		return res, fmt.Errorf("kvstore DEL revision error. expected %d, got %d", revision, responseRevision)
+	}
+	err = mctx.G().GetKVRevisionCache().MarkDeleted(mctx, entryID, revision)
+	if err != nil {
+		mctx.Debug("error loading delete of %+v into the revision cache: %v", entryID, err)
+		return res, err
+	}
+	return keybase1.KVDeleteEntryResult{
+		TeamName:  arg.TeamName,
+		Namespace: arg.Namespace,
+		EntryKey:  arg.EntryKey,
+		Revision:  revision,
+	}, nil
+}
+
+type getListNamespacesAPIRes struct {
+	Status     libkb.AppStatus `json:"status"`
+	TeamID     keybase1.TeamID `json:"team_id"`
+	Namespaces []string        `json:"namespaces"`
+}
+
+func (k *getListNamespacesAPIRes) GetAppStatus() *libkb.AppStatus {
+	return &k.Status
+}
+
+func (h *KVStoreHandler) ListKVNamespaces(ctx context.Context, arg keybase1.ListKVNamespacesArg) (res keybase1.KVListNamespaceResult, err error) {
+	ctx = libkb.WithLogTag(ctx, "KV")
+	mctx := libkb.NewMetaContext(ctx, h.G())
+	defer mctx.TraceTimed(fmt.Sprintf("KVStoreHandler#ListKVNamespaces: t:%s", arg.TeamName), func() error { return err })()
+	if err := h.assertLoggedIn(ctx); err != nil {
+		mctx.Debug("not logged in err: %v", err)
+		return res, err
+	}
+	teamID, err := h.resolveTeam(mctx, arg.TeamName)
+	if err != nil {
+		mctx.Debug("error resolving team with name %s: %v", arg.TeamName, err)
+		return res, err
+	}
+
+	var apiRes getListNamespacesAPIRes
+	apiArg := libkb.APIArg{
+		Endpoint:    "team/storage/list",
+		SessionType: libkb.APISessionTypeREQUIRED,
+		Args: libkb.HTTPArgs{
+			"team_id": libkb.S{Val: teamID.String()},
+		},
+	}
+	err = mctx.G().API.GetDecode(mctx, apiArg, &apiRes)
+	if err != nil {
+		return res, err
+	}
+	if apiRes.TeamID != teamID {
+		mctx.Debug("list KV Namespaces server returned an unexpected, mismatching teamID")
+		return res, fmt.Errorf("expected teamID %s from the server, got %s", teamID, apiRes.TeamID)
+	}
+	return keybase1.KVListNamespaceResult{
+		TeamName:   arg.TeamName,
+		Namespaces: apiRes.Namespaces,
+	}, nil
+}
+
+type getListEntriesAPIRes struct {
+	Status    libkb.AppStatus      `json:"status"`
+	TeamID    keybase1.TeamID      `json:"team_id"`
+	Namespace string               `json:"namespace"`
+	EntryKeys []compressedEntryKey `json:"entry_keys"`
+}
+
+type compressedEntryKey struct {
+	EntryKey string `json:"k"`
+	Revision int    `json:"r"`
+}
+
+func (k *getListEntriesAPIRes) GetAppStatus() *libkb.AppStatus {
+	return &k.Status
+}
+
+func (h *KVStoreHandler) ListKVEntries(ctx context.Context, arg keybase1.ListKVEntriesArg) (res keybase1.KVListEntryResult, err error) {
+	ctx = libkb.WithLogTag(ctx, "KV")
+	mctx := libkb.NewMetaContext(ctx, h.G())
+	defer mctx.TraceTimed(fmt.Sprintf("KVStoreHandler#ListKVEntries: t:%s, n:%s", arg.TeamName, arg.Namespace), func() error { return err })()
+	if err := h.assertLoggedIn(ctx); err != nil {
+		mctx.Debug("not logged in err: %v", err)
+		return res, err
+	}
+	teamID, err := h.resolveTeam(mctx, arg.TeamName)
```

Comment:
Doesn't have to be this PR. But this API should use team IDs instead of names. And maybe have one RPC to resolve a team name. That way it can be used for implicit teams without using the hidden names. And subteams without being susceptible to switching confusion. Chat is making this switch too.

Reply:
somewhere we need to resolve the team name to an ID unless we expect users to lookup and save their team ids. which seems gross. we could move this resolution up into the handler, but that doesn't seem like a big win to me, especially since the handlers are so much harder to test.

Reply:
I'm advocating for expecting users to lookup and save their team IDs. How would using implicit teams work?

Reply:
i don't think using `mlsteele,xgess` is so crazy. and if this doesn't work correctly with SBS, i think that's fine. i haven't really tested it. does SBS change the underlying team_id? the only place team name is used is at the layer closest to the user. 

i think passing team name is much nicer. we could pass back the team_id with each request and allow that parameter in addition to (and as an override to) team name. either way, yeah. let's discuss offline. it's not a hard change to make.